v-next/hardhat-build-system/src/internal/solidity/compiler/index.ts

```diff
@@ -17,66 +18,55 @@ export interface ICompiler {
 }
 
 export class Compiler implements ICompiler {
-  readonly #pathToSolcJs;
-  #loadedSolc?: any;
+  readonly #pathToSolcJs: string;
 
-  constructor(_pathToSolcJs: string) {
-    this.#pathToSolcJs = _pathToSolcJs;
+  constructor(pathToSolcJs: string) {
+    this.#pathToSolcJs = pathToSolcJs;
   }
 
   public async compile(input: CompilerInput) {
-    const solc = await this.getSolc();
+    const scriptPath = fileURLToPath(import.meta.resolve("./solcjs-runner.js"));
 
-    const jsonOutput = solc.compile(JSON.stringify(input));
-    return JSON.parse(jsonOutput);
-  }
-
-  public async getSolc() {
-    if (this.#loadedSolc !== undefined) {
-      return this.#loadedSolc;
-    }
-
-    const solcWrapper = require("solc/wrapper");
-    this.#loadedSolc = solcWrapper(
-      this.#loadCompilerSources(this.#pathToSolcJs),
-    );
-
-    return this.#loadedSolc;
-  }
-
-  /**
-   * This function loads the compiler sources bypassing any require hook.
-   *
-   * The compiler is a huge asm.js file, and using a simple require may trigger
-   * babel/register and hang the process.
-   */
-  #loadCompilerSources(compilerPath: string) {
-    /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions --
-    TODO: This was already replaced in `main`. */
-    const Module = module.constructor as any;
-
-    // if Hardhat is bundled (for example, in the vscode extension), then
-    // Module._extenions might be undefined. In that case, we just use a plain
-    // require.
-    if (Module._extensions === undefined) {
-      return require(compilerPath);
-    }
+    // If the script is a TypeScript file, we need to pass the --import tsx/esm
+    // which is available, as we are running the tests
+    const nodeOptions = scriptPath.endsWith(".ts")
+      ? ["--import", "tsx/esm"]
+      : [];
 
-    const previousHook = Module._extensions[".js"];
+    const output: string = await new Promise((resolve, reject) => {
+      try {
+        const subprocess = execFile(
+          process.execPath,
+          [...nodeOptions, scriptPath, this.#pathToSolcJs],
+          {
+            maxBuffer: 1024 * 1024 * 500,
```

Comment:
Could we add a short comment explaining why this specific memory size was chosen?

Reply:
TBH I copied that from the other compiler, so I don't really know 😅